### DISCUSSION TITLE
Character Virtual missing contact normal

### DISCUSSION THREAD

**Discussion by LeXXik** [jrouwe/JoltPhysics.js#188] [Q&A] [answered]
Trying to debug why the character virtual sometimes doesn't report a ground contact normal. It appears in my case that when the character moves on a slope uphill, the contact normal is received, but when it is going downhill, then there is no contact reported.

Initial debugging shows that when character moves downhill, `OnContactValidate` is called correctly, but the subsequent `OnContactSolve` is not. Any ideas why that could be or what is causing a validated contact not to process in solve? Could it be supporting volume being too high/low? Although, I'd suspect, it would behave same going up or down the slope, so must be something else.

**Comment by jrouwe** [comment]
Typically if you go downhill gravity is not strong enough for the character to stay connected to the ground and you'll float a little bit above the terrain before falling back onto it in subsequent frames. This is because a character has 0 vertical velocity if it is on the floor and usually a game character can accellerate very quickly, quicker than gravity accelerates in the vertical direction.

Are you using `CharacterVirtual::ExtendedUpdate` with a non zero `ExtendedUpdateSettings::mStickToFloorStepDown`? This mechanism tries to correct for that effect by forcefully moving the character back onto the floor if it is floating a little bit. Looking through the code, there may be a missing callback there too. If that's the case I'll fix it.

**Reply by jrouwe** [reply, marked as answer]
I've pushed a fix for CharacterVirtual StickToFloor so that it will trigger OnContactAdded callbacks. OnContactSolve will not be called, since that is only used when trying to find a sliding direction, which is not something that triggers during StickToFloor. It's also integrated in JoltPhysics.js now.

**Comment by LeXXik** [comment]
Thank you! Another question on the same topic. It looks like I can set the step height via the update settings and also via character virtual:

https://github.com/jrouwe/JoltPhysics.js/blob/e2b946880fa893cad08bb81fba4a95cf3daec5de/JoltJS.idl#L2750
https://github.com/jrouwe/JoltPhysics.js/blob/e2b946880fa893cad08bb81fba4a95cf3daec5de/JoltJS.idl#L2913

I guess the one on the character is more flexible, as it looks in 3 directions and can do filtering. I think I am safe to assume that I can use it instead of update settings in all cases. Is that correct?

Edit:
Actually, it looks like it might be used by the Character Virtual internally? As I don't think I need to track the forward direction of the character, etc. I guess defining the step heights and whether to stick to floor when going down should be controlled via the update settings then?

**Reply by jrouwe** [reply]
It is indeed the same thing. Calling `ExtendedUpdate` with the `ExtendedUpdateSettings` parameter passes that information on to `WalkStairs` internally.

`ExtendedUpdate` is nothing more than a call to `Update`, `WalkStairs` and `StickToFloor` in a particular order and with some glue code.

The forward direction is taken from the linear velocity.
